### ISSUE TITLE
Adding an optional 'lower' to the colum for case insensitive searches in the filter for PostgreSQL

### ISSUE THREAD

**Issue by kukel** [Kreyu/data-table-bundle#50]
Hey, 

We are using a PostgreSQL database which is case sensitive when doing a search. This can unfortunately not be changed and for our own queries we do something like 
```

$pattern = strtolower('HEllO WorlD');
$q = Doctrine_Query::create()
    ->select('u.username')
    ->from('User u')
    ->where("LOWER(u.username) LIKE ?", $pattern);
```

This works great and we have special indexes for the case insensitive compare. Is this something that could (selectively) be added? E.g. by setting an option on `StringFilterType` with a default of false. 

```
        $resolver
            ->setDefaults([
                'default_operator' => Operator::Contains,
                'supported_operators' => [
                    Operator::Equals,
                    Operator::NotEquals,
                    Operator::Contains,
                    Operator::NotContains,
                    Operator::StartsWith,
                    Operator::EndsWith,
                ],
                'toLower' => false,
            ])
```
And in a DataTableType on the builder do
```
 ->addFilter('name', StringFilterType::class, ['toLower' => true])
```
I unfortunately could not quickly find the best place or solution to create a pull request. Either solution I got to work means either pushing the options down into the FilterTypes or reworking the expression in the AbstractDoctrineOrmFilterType. Not sure what you'd prefer or maybe you have another solution (due to being more familiar with the code.

Thanks in advance

**Comment by Kreyu** [comment]
Hey @kukel!

Indeed that would be a useful feature. In the meantime, you can put the `LOWER` function in the filter `query_path` option:

```php
$builder->addFilter('name', StringFilterType::class, [
    'query_path' => 'LOWER(product.name)',
]);
```

**Note**: unless the [ `auto_alias_resolving`](https://data-table-bundle.swroblewski.pl/reference/filters/types/doctrine-orm/#auto_alias_resolving) option is set to `false`, `DoctrineOrmProxyQuery` will add root alias to the query path if it is not present. Unfortunately this means that setting `LOWER(name)` will result in `product.LOWER(name)`, since I haven't considered this case. This means the root alias should be set explicitly in the `query_path`, and can be retrieved from the query builder (for flexibility) like so:

```php
$rootAlias = current($builder->getQuery()->getQueryBuilder()->getRootAliases());

$builder->addFilter('name', StringFilterType::class, [
    'query_path' => "LOWER($rootAlias.name)",
]);
```

**Comment by Kreyu** [comment]
Hmm, one way would be adding a `FilterTypeExtension`, that would convert the filter data to lowercase like so:

```php
<?php

declare(strict_types=1);

namespace App\DataTable\Filter\Extension;

use Kreyu\Bundle\DataTableBundle\Filter\Extension\AbstractFilterTypeExtension;
use Kreyu\Bundle\DataTableBundle\Filter\FilterBuilderInterface;
use Kreyu\Bundle\DataTableBundle\Filter\FilterData;
use Kreyu\Bundle\DataTableBundle\Filter\FilterInterface;
use Kreyu\Bundle\DataTableBundle\Filter\Type\FilterType;
use Kreyu\Bundle\DataTableBundle\Query\ProxyQueryInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class LowercaseFilterTypeExtension extends AbstractFilterTypeExtension
{
    public function apply(ProxyQueryInterface $query, FilterData $data, FilterInterface $filter, array $options): void
    {
        if (!$options['to_lower']) {
            return;
        }

        if (is_string($value = $data->getValue())) {
            $data->setValue(mb_strtolower($value));
        }
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver
            ->setDefault('to_lower', false)
        ;
    }

    public static function getExtendedTypes(): iterable
    {
        return [FilterType::class];
    }
}
```

There's one side effect though- the active filter "clear" button will also show lowercased value. I think applying the lowercased value **but** still displaying the original one would require tinkering in the built-in Doctrine ORM filter classes - extending them to add `mb_strtolower` to the `getFilterValue()`

**Comment by asdfMaciej** [comment]
Hey @kukel - will this work for you?

```php
<?php

declare(strict_types=1);

namespace App\DataTable\Filter;

use Kreyu\Bundle\DataTableBundle\Bridge\Doctrine\Orm\Filter\Type\AbstractFilterType;
use Kreyu\Bundle\DataTableBundle\Bridge\Doctrine\Orm\Query\DoctrineOrmProxyQuery;
use Kreyu\Bundle\DataTableBundle\Filter\FilterData;
use Kreyu\Bundle\DataTableBundle\Filter\FilterInterface;
use Kreyu\Bundle\DataTableBundle\Filter\Operator;
use Kreyu\Bundle\DataTableBundle\Query\ProxyQueryInterface;
use Symfony\Component\Form\Extension\Core\Type\TextType;
use Symfony\Component\OptionsResolver\OptionsResolver;

class LowerFilterType extends AbstractFilterType
{
    /**
     * @param DoctrineOrmProxyQuery $query
     */
    public function apply(ProxyQueryInterface $query, FilterData $data, FilterInterface $filter, array $options): void
    {
        $value = $data->getValue();
        $parameterName = $this->getUniqueParameterName($query, $filter);

        $expression = $query->expr()->like(
            'LOWER('.$this->getFilterQueryPath($query, $filter).')',
            "LOWER(:$parameterName)",
        );

        $query
            ->andWhere($expression)
            ->setParameter($parameterName, '%'.$value.'%');
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefault('field_type', TextType::class);
        $resolver->setDefault('operator_options', function (OptionsResolver $resolver) {
            $resolver->setDefaults([
                'visible' => false,
            ]);
        });
    }
}
```

**Comment by Kreyu** [comment]
FYI: ability to transform the expression (applying the LOWER, UPPER, TRIM and custom functions) will be supported in the Doctrine ORM integration bundle in the next version (#54).

**Comment by Kreyu** [comment]
Doctrine ORM filters now support expression transformers and `upper`, `lower` and `trim` options (version 0.16). For more information, [see docs](https://data-table-bundle.swroblewski.pl/docs/integrations/doctrine-orm/expression-transformers.html).

TLDR;

```php
$builder->addFilter('name', StringFilterType::class, [
    'trim' => true,
]);
```

will result in something like:

![image](https://github.com/Kreyu/data-table-bundle/assets/24738141/a1edaec3-d632-410c-bafc-1b90fab576bf)

same with `lower` and `upper` options.
